npep/npep-122.md

```diff
@@ -0,0 +1,194 @@
+from accidentally interfering with each other, but still want to allow them to communicate if needed. 
+
+In reality, the CR attached to this story defines strict tenancy, which should be a separate use case (see next section).
+
+To make the use case more obvious we can add some details to it like the following
+
+    As a cluster admin, I want to build tenants in my cluster that are isolated from each other by default.
+    This should prevent tenants from interfering with each other by accident, but may be overridden by namespace
+    owners when needed.
+    Tenancy may be modeled as 1:1, where 1 tenant is mapped to a single Namespace, or 1:n, where a single tenant 
+    may own more than 1 Namespace.
+
+#### Story 4.2: Create and Isolate multiple tenants in a cluster, strict isolation
+
+Strict tenancy is the "Coke vs Pepsi" sort of thing where you want each tenant to feel like it has its own cluster, 
+and be totally independent of the other tenants. We can write it down like this
+
+    As a cluster admin, I want to build tenants in my cluster that are isolated from each other and this can't
+    be overridden. This policy should make every tenant completely independent and isolated from another tenants.
+    Tenancy may be modeled as 1:1, where 1 tenant is mapped to a single Namespace, or 1:n, where a single tenant 
+    may own more than 1 Namespace.
+
+#### Story 4.3: Allow internal connections for tenants
+    
+    As a cluster admin, I want to build tenants in my cluster and always allow connections inside one tenant.
+    At the same time I want to setup an overridable deny-all policy to protect namespaces by default.
+    This policy should make sure internal connectivity for a tenant is always allowed, in case there are
+    lower-priority deny rules.
+
+#### Story 4.4: Tenants interaction with (B)ANP
+
+    As a cluster admin, I want to be able to setup policies with higher and lower priority than tenancy policy.
+    Higher priority to allow/deny some inter-tenant connections, for example, allow ingress traffic from the
+    monitoring namespace without including that namespace to every tenant. Lower priority to allow/deny
+    connections that are not a part of tenant policy, for example, allow internal tenant connections with tenancy
+    and deny-all policy with lower priority.
+
+#### Story 4.5: Ensure every namespace in the cluster is a part of some tenant
+
+    As a cluster admin, I want to implement security zones tenants, and deny connections between different zones.
+    I want to use a label e.g. `security-zone` to build tenants, and I want to make sure every namespace in the cluster
+    is a part of some tenant/security zone. This way, I can be sure that even if the `security-zone` 
+    label is not assigned to the tenant for some reason, tenancy rules will still apply and security zones won't be
+    violated.
+
+#### What I couldn't figure our user stories for
+
+- Skip action
```

Comment:
hrm I feel like a skip action could help with 4.3 and 4.4 

"overridable deny-all policy to protect namespaces by default."

Could be implemented with a ANP + pass rules OR a BANP 

IMO at the end of the day it's hard because we're thinking about implementation details when writing user stories 🤣  even though a story can be a bit open ended leaving the design up to the implementer

Reply:
> "overridable deny-all policy to protect namespaces by default."
> 
> Could be implemented with a ANP + pass rules OR a BANP

How would you do overridable deny-all with Pass? If the rule Denies then it's not overridable and if it Passes then it's not denying

Reply:
I agree we shouldn't focus too much on this, but since I did spend some time trying to figure out a use case for it and couldn't, I decided to write it down :)
Skip action is mostly used to optimize interaction between different ANPs (if we consider tenancy and ANP to be different things in this context), so that it allows to ignore the ANPs with lower-level priorities (ANP<->ANP relations). We know that every skip action can be replaced with just not selecting Skipped pods in the netpols with lower prio, but skipping allows more flexibility. From that perspective, I don't think it makes sense for tenancy to define a skip action for other ANPs (tenancy<->ANP), same as I don't think there is a reasonable use case to skip tenancy with the lower priority (tenancy<->tenancy).
lmk if that makes any sense :)

Reply:
Please take a look at the second example yaml in https://network-policy-api.sigs.k8s.io/reference/examples/#sample-spec-for-story-4-create-and-isolate-multiple-tenants-in-a-cluster. This is how we originally envisioned `Pass` is used in creating tenancy rules

Reply:
very nice link, thank you @Dyanngg! 
I am not sure I understand that example: why would you write 
```
apiVersion: policy.networking.k8s.io/v1alpha1
kind: AdminNetworkPolicy
metadata:
  name: tenant-creation-example
spec:
  priority: 50
  subject:
    namespaces:
      matchExpressions: {key: "tenant"; operator: Exists}
  ingress:
    - action: Pass # Pass inter-tenant traffic to any defined NetworkPolicies
      from:
      - namespaces:
          sameLabels:
          - tenant
    - action: Deny   # Deny everything else other than same tenant traffic
      from:
      - namespaces:
          namespaceSelector: {}
```
when it is equivalent to
```
apiVersion: policy.networking.k8s.io/v1alpha1
kind: AdminNetworkPolicy
metadata:
  name: tenant-creation-example
spec:
  priority: 50
  subject:
    namespaces:
      matchExpressions: {key: "tenant"; operator: Exists}
  ingress:
    - action: Deny   # Deny everything else other than same tenant traffic
      from:
      - namespaces:
          namespaceSelector: {}
```
and doesn't seem to actually do what the user story wanted? Maybe I am missing something here?

Reply:
I think I see @Dyanngg's point -- by putting in an explicit `Pass` directive, you make the rule more precise.

quick nit: I think if we want to drop the pass, the rule shuold be:
```
apiVersion: policy.networking.k8s.io/v1alpha1
kind: AdminNetworkPolicy
metadata:
  name: tenant-creation-example
spec:
  priority: 50
  subject:
    namespaces:
      matchExpressions: {key: "tenant"; operator: Exists}
  ingress:
    - action: Deny   # Deny everything else other than same tenant traffic
      from:
      - namespaces:
          notSameLabels:
          - tenant
```

While the example you provide does indeed achieve the same in isolation, what happens if you have multiple rules. For example a rule at priority 100 (evaluated after the rule at priority 50) could allow traffic between pods with the label `my-app: example`. By having the pass rule at priority 50, they're able to say "I've defined a tenancy relationship at rule 50. Inter-tenant traffic is dropped, and traffic within a tenant is up to the namespace admins. Now for anything not selected by that tenancy rule, allow traffic between apps with the label `my-app: example`"

While the toy example of `my-app: example` is a little contrived, a cluster admin shouldn't have to always put in an exception if they want to provide a rule with lower precedence than a tenancy rule (i.e. rule at 100 needs to explicitly not select the tenant traffic).

Reply:
@rahulkjoshi your yaml is absolutely correct, and it is actually listed as the first option https://network-policy-api.sigs.k8s.io/reference/examples/#sample-spec-for-story-4-create-and-isolate-multiple-tenants-in-a-cluster
so the question I have is why we need the second option, because I am trying to figure out if Pass action makes sense for tenancy in general, and in this second option yaml in particular

Reply:
I think this is a good section leave... however on the API side of things I think it's alright to still possibly use pass in whatever we come up with during the design stage.... remember we've agreed that it's OK not to have an explicit user story for re-applying a feature that's already in the api, However, if we chose to DISALLOW pass rules when building tenants then we will need some explicit documentation that describes why.

Reply:
exactly, this section is here just to be explicit on the use case, and doesn't imply anything for the API itself